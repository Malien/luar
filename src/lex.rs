use logos::{Lexer, Logos};
use thiserror::Error;

/// Syntax
///
/// ```
/// module -> { statement | function }
/// block -> { stat sc } [ret sc]
/// sc -> ';'
/// stat -> varlist1 '=' explist1
/// varlist1 -> var { ',' var }
/// var -> name | var '[' exp1 ']' | var '.' name
/// stat -> 'while' exp1 'do' block 'end'
///       | 'repeat' block 'until' exp1
///       | 'if' exp1 'then' block { elseif } ['else' block] 'end'
///       | functioncall
///       | tableconstructor
///       | 'local' declist [init]
/// elseif -> 'elseif' exp1 'then' block
/// ret -> 'return' explist
/// declist -> name { , 'name' }
/// init -> '=' explist1
/// exp -> '(' exp ')'
///      | 'nil'
///      | number
///      | literal
///      | var
/// tableconstructor -> '@' '(' [exp1] ')' | '@' [name] fieldlist
/// fieldlist -> '{' [ffieldlist1] '}' | '[' [lfieldlist1] ']'
/// ffieldlist1 -> ffield { ',' ffield }
/// ffield -> name '=' exp
/// lfieldlist1 -> exp { ',' exp }
/// functioncall -> var '(' [explist1] ')'
/// explist1 -> { exp1 ',' } exp
/// function -> 'function' name '(' [parlist1] ')' block 'end'
/// parlist1 -> 'name' { ',' name }
/// ```

/// Reserved words:
///     `and` `do` `else` `elseif` `end` `function` `if` `local` `nil` `not`
///     `or` `repeat` `return` `until` `then` `while`
///
/// Other tokens:
///     `~=` `<=` `>=` `<` `>` `=` `..` `+` `-` `*` `/` `%` `(` `)` `{` `}` `[` `]` `@` `;` `,` `.`
///
/// Comments denoted by `--` and continue until the end of the line
///
/// Strings are delimited with either single or double quotes, and can contain C-like escape sequences (like `\n` `\r` `\t`)
///
/// Numbers are have either
///     `4` `4.23` `4.` `.23` `4.57e-7` `.3e4`

// #[derive(Copy, Clone, Debug, Eq, PartialEq, Logos)]
// enum Token {
//
// }

#[derive(Clone, Debug, PartialEq, Eq, Logos)]
pub enum Token {
    #[error]
    #[regex(r"[ \t\n\f]", logos::skip)]
    Error,
    #[token("and")]
    And,
    #[token("do")]
    Do,
    #[token("else")]
    Else,
    #[token("elseif")]
    ElseIf,
    #[token("end")]
    End,
    #[token("function")]
    Function,
    #[token("if")]
    If,
    #[token("local")]
    Local,
    #[token("nil")]
    Nil,
    #[token("not")]
    Not,
    #[token("or")]
    Or,
    #[token("repeat")]
    Repeat,
    #[token("return")]
    Return,
    #[token("until")]
    Until,
    #[token("then")]
    Then,
    #[token("while")]
    While,
    #[token("~=")]
    NotEquals,
    #[token("<=")]
    LessOrEquals,
    #[token(">=")]
    GreaterOrEquals,
    #[token(">")]
    Greater,
    #[token("<")]
    Lesser,
    #[token("=")]
    Equals,
    #[token("..")]
    Concat,
    #[token("+")]
    Plus,
    #[token("-")]
    Minus,
    #[token("*")]
    Multiply,
    #[token("/")]
    Div,
    #[token("%")]
    Mod,
    #[token("(")]
    OpenRoundBracket,
    #[token(")")]
    CloseRoundBracket,
    #[token("[")]
    OpenSquareBracket,
    #[token("]")]
    CloseSquareBracket,
    #[token("{")]
    OpenSquigglyBracket,
    #[token("}")]
    CloseSquigglyBracket,
    #[token("@")]
    At,
    #[token(".")]
    Dot,
    #[token(",")]
    Comma,
    #[token(";")]
    Semicolon,
    #[regex(r"[_a-zA-Z][_a-zA-Z0-9]*", make_owned_string)]
    Ident(String),
    #[regex(
        "(\"(?:[^\"'\\\\]|\\\\.)*\")|('(?:[^\"'\\\\]|\\\\.)*')",
        parse_string_literal
    )]
    StringLiteral(String),
    #[regex(
        r"[+-]?((\d+\.\d+)|(\.\d+)|(\d+\.?))(e[+-]?\d+)?",
        parse_number_literal
    )]
    Number(NumberLiteral),
}

#[derive(Error, Debug)]
#[error("Token passed in is not a valid string literal")]
struct StringLiteralParseError;

// enum NumberLiteralParseError {
//     Integral()
// }
type NumberLiteralParseError = num::rational::ParseRatioError;

// Can be optimized not to allocate extra space if not necessary
// enum LuaNumberLiteral {
//     Integral(BigInt),
//     Rational(BigRational)
// }
type NumberLiteral = num::rational::BigRational;

fn parse_string_literal(input: &mut Lexer<Token>) -> Result<String, StringLiteralParseError> {
    let slice = input.slice();
    if slice.len() < 2 {
        return Err(StringLiteralParseError);
    }
    let raw_string: &str = &slice[1..slice.len() - 1];
    // Not a very efficient method to escape characters
    Ok(raw_string
        .replace(r"\\", "\\")
        .replace(r"\r", "\r")
        .replace(r"\n", "\n")
        .replace(r"\t", "\t")
        .replace("\\\"", "\"")
        .replace(r"\'", "'"))
}

// static integer_regex: Regex = Regex::new(r"[+-]?(\d+\.?)(e[+-]?\d+)?").unwrap();

fn parse_number_literal(
    input: &mut Lexer<Token>,
) -> Result<NumberLiteral, NumberLiteralParseError> {
    let slice: &str = input.slice();
    slice.parse()
}

fn make_owned_string(input: &mut Lexer<Token>) -> String {
    input.slice().to_string()
}

#[cfg(test)]
mod tests {
    use super::Token;
    use logos::Logos;

    macro_rules! assert_tokens {
        ($name:ident, $text:tt) => {
            #[test]
            #[function_name::named]
            fn $name() {
                static TEXT: &str = indoc::indoc!($text);
                let tokens: Vec<Token> = Token::lexer(TEXT).into_iter().collect();
                insta::assert_debug_snapshot!(function_name!(), tokens);
            }
        };
    }

    assert_tokens!(function_1, "
        function remove_blanks (s)
            local b = strfind(s, ' ')
            while b do
                s = strsub(s, 1, b-1) .. strsub(s, b+1)
                b = strfind(s, ' ')
            end
            return s
        end
    ");
}
