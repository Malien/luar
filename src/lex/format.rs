use super::Token;

pub struct Formatting {
    pub before: FormattingStyle,
    pub after: FormattingStyle,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
pub enum FormattingStyle {
    Condensed,
    Space,
    StrictSpace,
    Newline,
    Indent(IndentationChange),
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
#[repr(i8)]
pub enum IndentationChange {
    Increase = 1,
    Decrease = -1,
}

pub fn formatting_style_precedence(
    left: FormattingStyle,
    right: FormattingStyle,
) -> FormattingStyle {
    use FormattingStyle::*;
    use IndentationChange::*;

    match (left, right) {
        (Condensed, Space) => Condensed,
        (Space, Condensed) => Condensed,

        (Condensed, StrictSpace) => StrictSpace,
        (StrictSpace, Condensed) => StrictSpace,

        (Indent(Increase), Indent(Decrease)) => Newline,
        (Indent(c), _) => Indent(c),
        (_, Indent(c)) => Indent(c),

        (Newline, _) => Newline,
        (_, Newline) => Newline,

        // when a == b
        (a, _) => a,
    }
}

impl Token {
    pub fn formatting(&self) -> Formatting {
        use FormattingStyle::*;
        use IndentationChange::*;
        use Token::*;
        match self {
            Error | And | Nil | Not | Or | Equals | NotEquals | LessOrEquals | GreaterOrEquals
            | Greater | Less | Concat | Plus | Minus | Mul | Div | Mod | OpenSquigglyBracket
            | CloseSquigglyBracket | Ident(_) | String(_) | Number(_) => Formatting {
                before: Space,
                after: Space,
            },
            Function | Local => Formatting {
                before: Newline,
                after: Space,
            },
            If | Return | While => Formatting {
                before: Newline,
                after: StrictSpace,
            },
            Else => Formatting {
                before: Indent(Decrease),
                after: Indent(Increase),
            },
            End => Formatting {
                before: Indent(Decrease),
                after: Newline,
            },
            Until | ElseIf => Formatting {
                before: Indent(Decrease),
                after: Space,
            },
            Then | Do | Repeat => Formatting {
                before: Space,
                after: Indent(Increase),
            },
            Exp | OpenRoundBracket | OpenSquareBracket
            | Dot | Colon => Formatting {
                before: Condensed,
                after: Condensed,
            },
            Comma | Semicolon | CloseRoundBracket | CloseSquareBracket => Formatting {
                before: Condensed,
                after: Space,
            },
            Assignment => Formatting {
                before: StrictSpace,
                after: StrictSpace,
            },
        }
    }
}

pub fn format_single_token(
    token: Token,
    indent: &mut i32,
    current_format: &mut FormattingStyle,
    buf: &mut dyn std::fmt::Write,
) -> std::fmt::Result {
    use FormattingStyle::*;
    let Formatting { before, after } = token.formatting();
    match formatting_style_precedence(*current_format, before) {
        Condensed => {}
        Space | StrictSpace => buf.write_char(' ')?,
        Newline => {
            buf.write_char('\n')?;
            for _ in 0..*indent {
                buf.write_char('\t')?;
            }
        }
        Indent(change) => {
            buf.write_char('\n')?;
            *indent += change as i32;
            for _ in 0..*indent {
                buf.write_char('\t')?;
            }
        }
    };
    *current_format = after;
    write!(buf, "{}", token)?;
    Ok(())
}

pub fn format_tokens(
    tokens: &mut impl Iterator<Item = Token>,
    buf: &mut dyn std::fmt::Write,
) -> std::fmt::Result {
    if let Some(first_token) = tokens.next() {
        let mut indent = 0;
        let Formatting {
            after: mut current_format,
            ..
        } = first_token.formatting();
        write!(buf, "{}", first_token)?;

        for token in tokens {
            format_single_token(token, &mut indent, &mut current_format, buf)?;
        }
    }
    Ok(())
}

#[macro_export]
macro_rules! fmt_tokens {
    ($type:ty) => {
        impl std::fmt::Display for $type {
            fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
                crate::lex::format::format_tokens(&mut self.clone().to_tokens(), fmt)
            }
        }
    };
}
