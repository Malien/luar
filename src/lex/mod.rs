mod ident;
mod number_literal;
mod string_literal;
mod token;

pub type DynTokens = Box<dyn Iterator<Item = Token>>;

pub trait ToTokenStream {
    type Tokens: IntoIterator<Item = Token>;

    fn to_tokens(self) -> Self::Tokens;
}

#[macro_export]
macro_rules! fmt_tokens {
    ($type:ty) => {
        impl std::fmt::Display for $type {
            fn fmt(&self, fmt: &mut std::fmt::Formatter) -> std::fmt::Result {
                for token in self.clone().to_tokens() {
                    write!(fmt, "{} ", token)?;
                }
                Ok(())
            }
        }
    };
}

pub use ident::Ident;
pub use number_literal::NumberLiteral;
pub use string_literal::StringLiteral;
pub use token::Token;
