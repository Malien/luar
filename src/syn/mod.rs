use crate::lex::{Ident, NumberLiteral, StringLiteral, Token};

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum BinaryOperator {
    // Precedence level 0
    And,
    Or,
    // Precedence level 1
    Less,
    Greater,
    LessOrEquals,
    GreaterOrEquals,
    NotEquals,
    Equals,
    // Precedence level 2
    Concat,
    // Precedence level 3
    Plus,
    Minus,
    // Precedence level 4
    Mul,
    Div,
    // Precedence level 5
    Exp,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum UnaryOperator {
    Minus,
    Not,
}

#[derive(Debug, PartialEq, Eq, Clone, Copy)]
pub enum TableConstructor {}

#[derive(Debug, PartialEq, Clone)]
pub enum Expression {
    Nil,
    String(StringLiteral),
    Number(NumberLiteral),
    Variable(Var),
    BinaryOperator {
        lhs: Box<Expression>,
        op: BinaryOperator,
        rhs: Box<Expression>,
    },
    UnaryOperator {
        op: UnaryOperator,
        exp: Box<Expression>,
    },
    TableConstructor(TableConstructor),
    FunctionCall {
        func: Var,
        args: Vec<Expression>,
    },
}

#[derive(Debug, PartialEq, Clone)]
pub enum Var {
    Named(Ident),
    PropertyAccess {
        from: Box<Var>,
        property: Ident,
    },
    MemberLookup {
        from: Box<Var>,
        value: Box<Expression>,
    },
}

enum VarLeftover {
    Nothing,
    PropertyAccess {
        from: Box<VarLeftover>,
        property: Ident,
    },
    MemberLookup {
        from: Box<VarLeftover>,
        value: Expression,
    },
}

fn accumulate_var_leftovers(base: Var, leftovers: VarLeftover) -> Var {
    match leftovers {
        VarLeftover::Nothing => base,
        VarLeftover::PropertyAccess { from, property } => accumulate_var_leftovers(
            Var::PropertyAccess {
                from: Box::new(base),
                property,
            },
            *from,
        ),
        VarLeftover::MemberLookup { from, value } => accumulate_var_leftovers(
            Var::MemberLookup {
                from: Box::new(base),
                value: Box::new(value),
            },
            *from,
        ),
    }
}

// macro_rules! l_binary_op_expr {
//     ($token:pat, $op:expr) => {
//         lhs:(@) _:[$token] rhs:@ {
//             Expression::BinaryOperator {
//                 op: $op,
//                 lhs: Box::new(lhs),
//                 rhs: Box::new(rhs),
//             }
//         }
//     };
// }

// macro_rules! r_binary_op_expr {
//     ($token:pat, $op:expr) => {
//         lhs:@ _:[$token] rhs:(@) {
//             Expression::BinaryOperator {
//                 op: $op,
//                 lhs: Box::new(lhs),
//                 rhs: Box::new(rhs),
//             }
//         }
//     };
// }

peg::parser! {
    grammar lua_parser() for [Token] {
        pub rule nil() -> Expression
            = _:[Token::Nil] { Expression::Nil }

        pub rule string() -> Expression
            = _:[Token::String(literal)] { Expression::String(literal) }

        pub rule number() -> Expression
            = _:[Token::Number(literal)] { Expression::Number(literal) }

        pub rule var_expression() -> Expression
            = var:var() { Expression::Variable(var) }

        pub rule expression() -> Expression = precedence! {
            x:(@) _:[Token::And] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::And,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::Or] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Or,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            --
            x:(@) _:[Token::Lesser] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Less,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::Greater] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Greater,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::LessOrEquals] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::LessOrEquals,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::GreaterOrEquals] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::GreaterOrEquals,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::NotEquals] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::NotEquals,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::Equals] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Equals,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            --
            x:(@) _:[Token::Concat] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Concat,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            --
            x:(@) _:[Token::Plus] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Plus,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::Minus] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Minus,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            --
            x:(@) _:[Token::Multiply] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Mul,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            x:(@) _:[Token::Div] y:@ {
                Expression::BinaryOperator {
                    op: BinaryOperator::Div,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            --
            _:[Token::Not] x:@ {
                Expression::UnaryOperator {
                    op: UnaryOperator::Not,
                    exp: Box::new(x),
                }
            }
            _:[Token::Minus] x:@ {
                Expression::UnaryOperator {
                    op: UnaryOperator::Minus,
                    exp: Box::new(x),
                }
            }
            --
            x:@ _:[Token::Exp] y:(@) {
                Expression::BinaryOperator {
                    op: BinaryOperator::Exp,
                    lhs: Box::new(x),
                    rhs: Box::new(y),
                }
            }
            --
            e:nil() { e }
            e:string() { e }
            e:number() { e }
            e:var_expression() { e }
            _: [Token::OpenRoundBracket] e:expression() _:[Token::CloseRoundBracket] { e }
        }

        pub rule named() -> Var
            = _:[Token::Ident(ident)] { Var::Named(ident) }

        pub rule property_access() -> Var
            = base:var() _:[Token::Dot] _:[Token::Ident(property) ] {
                Var::PropertyAccess {
                    from: Box::new(base),
                    property
                }
            }

        pub rule var() -> Var
            = base:named() leftovers:_var() { accumulate_var_leftovers(base, leftovers) }

        rule _var() -> VarLeftover
            = _:[Token::Dot] _:[Token::Ident(ident)] next:_var() {
                VarLeftover::PropertyAccess {
                    from: Box::new(next),
                    property: ident
                }
            }
            / _:[Token::OpenSquareBracket] e:expression() _:[Token::CloseSquareBracket] next:_var() {
                VarLeftover::MemberLookup {
                    from: Box::new(next),
                    value: e
                }
            }
            / { VarLeftover::Nothing }
    }
}

#[cfg(test)]
mod tests {
    use indoc::indoc;
    use logos::Logos;
    use quickcheck::{Arbitrary, Gen};

    use super::{lua_parser, BinaryOperator, Expression, UnaryOperator, Var};
    use crate::lex::{Ident, NumberLiteral, StringLiteral, Token};

    use std::iter;

    #[derive(Debug, Clone)]
    struct ArbitraryTokens<T> {
        tokens: Vec<Token>,
        expected: T,
    }

    impl<T> From<(Vec<Token>, T)> for ArbitraryTokens<T> {
        fn from((tokens, expected): (Vec<Token>, T)) -> Self {
            Self { tokens, expected }
        }
    }

    impl Arbitrary for ArbitraryTokens<UnaryOperator> {
        fn arbitrary(g: &mut Gen) -> Self {
            match u8::arbitrary(g) % 2 {
                0 => (vec![Token::Minus], UnaryOperator::Minus),
                1 => (vec![Token::Not], UnaryOperator::Not),
                _ => unreachable!(),
            }
            .into()
        }
    }

    impl Arbitrary for ArbitraryTokens<BinaryOperator> {
        fn arbitrary(g: &mut Gen) -> Self {
            match u8::arbitrary(g) % 14 {
                0 => (vec![Token::And], BinaryOperator::And),
                1 => (vec![Token::Or], BinaryOperator::Or),
                2 => (vec![Token::Lesser], BinaryOperator::Less),
                3 => (vec![Token::Greater], BinaryOperator::Greater),
                4 => (vec![Token::LessOrEquals], BinaryOperator::LessOrEquals),
                5 => (
                    vec![Token::GreaterOrEquals],
                    BinaryOperator::GreaterOrEquals,
                ),
                6 => (vec![Token::NotEquals], BinaryOperator::NotEquals),
                7 => (vec![Token::Equals], BinaryOperator::Equals),
                8 => (vec![Token::Concat], BinaryOperator::Concat),
                9 => (vec![Token::Plus], BinaryOperator::Plus),
                10 => (vec![Token::Minus], BinaryOperator::Minus),
                11 => (vec![Token::Multiply], BinaryOperator::Mul),
                12 => (vec![Token::Div], BinaryOperator::Div),
                13 => (vec![Token::Exp], BinaryOperator::Exp),
                _ => unreachable!(),
            }
            .into()
        }
    }

    impl Arbitrary for ArbitraryTokens<Expression> {
        fn arbitrary(g: &mut Gen) -> Self {
            match u8::arbitrary(g) % 6 {
                0 => ArbitraryTokens {
                    tokens: vec![Token::Nil],
                    expected: Expression::Nil,
                },
                1 => {
                    let literal = NumberLiteral::arbitrary(g);
                    ArbitraryTokens {
                        tokens: vec![Token::Number(literal)],
                        expected: Expression::Number(literal),
                    }
                }
                2 => {
                    let literal = StringLiteral::arbitrary(g);
                    ArbitraryTokens {
                        tokens: vec![Token::String(literal.clone())],
                        expected: Expression::String(literal),
                    }
                }
                3 => {
                    let ArbitraryTokens { expected, tokens } = ArbitraryTokens::arbitrary(g);
                    ArbitraryTokens {
                        tokens,
                        expected: Expression::Variable(expected),
                    }
                }
                4 => {
                    let ArbitraryTokens {
                        expected: op,
                        tokens: op_tokens,
                    } = ArbitraryTokens::arbitrary(g);
                    let ArbitraryTokens {
                        expected: exp,
                        tokens: exp_tokens,
                    } = ArbitraryTokens::arbitrary(g);

                    let tokens: Vec<_> = iter::once(Token::OpenRoundBracket)
                        .chain(op_tokens)
                        .chain(exp_tokens)
                        .chain(iter::once(Token::CloseRoundBracket))
                        .collect();

                    ArbitraryTokens {
                        tokens,
                        expected: Expression::UnaryOperator {
                            op,
                            exp: Box::new(exp),
                        },
                    }
                }
                5 => {
                    let ArbitraryTokens {
                        expected: op,
                        tokens: op_tokens,
                    } = ArbitraryTokens::arbitrary(g);
                    let ArbitraryTokens {
                        expected: lhs,
                        tokens: lhs_tokens,
                    } = ArbitraryTokens::arbitrary(g);
                    let ArbitraryTokens {
                        expected: rhs,
                        tokens: rhs_tokens,
                    } = ArbitraryTokens::arbitrary(g);

                    let tokens: Vec<_> = iter::once(Token::OpenRoundBracket)
                        .chain(lhs_tokens)
                        .chain(op_tokens)
                        .chain(rhs_tokens)
                        .chain(iter::once(Token::CloseRoundBracket))
                        .collect();

                    ArbitraryTokens {
                        tokens,
                        expected: Expression::BinaryOperator {
                            op,
                            lhs: Box::new(lhs),
                            rhs: Box::new(rhs),
                        },
                    }
                }
                _ => unreachable!(),
            }
        }
    }

    #[test]
    fn nill_expr() {
        let parsed = lua_parser::expression(&[Token::Nil]).unwrap();
        assert_eq!(Expression::Nil, parsed);
    }

    #[quickcheck]
    fn number_expr(literal: NumberLiteral) {
        let expression = lua_parser::expression(&[Token::Number(literal)]).unwrap();
        match (&literal, &expression) {
            (NumberLiteral(x), Expression::Number(NumberLiteral(y))) if f64::is_nan(*x) => {
                assert!(f64::is_nan(*y))
            }
            _ => assert_eq!(Expression::Number(literal), expression),
        };
    }

    #[quickcheck]
    fn string_expr(literal: StringLiteral) {
        assert_eq!(
            Expression::String(literal.clone()),
            lua_parser::expression(&[Token::String(literal)]).unwrap()
        );
    }

    #[quickcheck]
    fn var_expr(ArbitraryTokens { tokens, expected }: ArbitraryTokens<Var>) {
        let parsed = lua_parser::expression(&tokens).unwrap();
        assert_eq!(parsed, Expression::Variable(expected));
    }

    #[quickcheck]
    fn parses_arbitrary_expression(
        ArbitraryTokens { tokens, expected }: ArbitraryTokens<Expression>,
    ) {
        let parsed = lua_parser::expression(&tokens).unwrap();
        assert_eq!(parsed, expected);
    }

    impl Arbitrary for ArbitraryTokens<Var> {
        fn arbitrary(g: &mut Gen) -> Self {
            if g.size() <= 1 {
                let ident = Ident::arbitrary(g);
                return (vec![Token::Ident(ident.clone())], Var::Named(ident)).into();
            }
            match u8::arbitrary(g) % 3 {
                0 => {
                    let ident = Ident::arbitrary(g);
                    (vec![Token::Ident(ident.clone())], Var::Named(ident)).into()
                }
                1 => {
                    let ident = Ident::arbitrary(g);
                    let ArbitraryTokens {
                        mut tokens,
                        expected,
                    } = ArbitraryTokens::arbitrary(g);
                    tokens.push(Token::Dot);
                    tokens.push(Token::Ident(ident.clone()));
                    (
                        tokens,
                        Var::PropertyAccess {
                            from: Box::new(expected),
                            property: ident,
                        },
                    )
                }
                2 => {
                    let expression = ArbitraryTokens::arbitrary(g);
                    let ArbitraryTokens {
                        mut tokens,
                        expected,
                    } = ArbitraryTokens::arbitrary(g);
                    tokens.push(Token::OpenSquareBracket);
                    tokens.extend(expression.tokens);
                    tokens.push(Token::CloseSquareBracket);
                    (
                        tokens,
                        Var::MemberLookup {
                            from: Box::new(expected),
                            value: Box::new(expression.expected),
                        },
                    )
                }
                _ => unreachable!(),
            }
            .into()
        }
    }

    #[quickcheck]
    fn parse_named_var(ident: Ident) {
        let parsed = lua_parser::var(&[Token::Ident(ident.clone())]).unwrap();
        assert_eq!(Var::Named(ident), parsed);
    }

    #[quickcheck]
    fn parse_single_ppty_access(base: Ident, property: Ident) {
        let parsed = lua_parser::var(&[
            Token::Ident(base.clone()),
            Token::Dot,
            Token::Ident(property.clone()),
        ])
        .unwrap();
        assert_eq!(
            Var::PropertyAccess {
                from: Box::new(Var::Named(base)),
                property: property
            },
            parsed
        );
    }

    #[quickcheck]
    fn parse_arbitrary_ppty_access(base: Ident, properties: Vec<Ident>) {
        let mut sequence = Vec::with_capacity(properties.len() + 1);
        sequence.push(Token::Ident(base.clone()));
        sequence.extend(properties.iter().cloned().flat_map(|property| {
            std::iter::once(Token::Dot).chain(std::iter::once(Token::Ident(property)))
        }));
        let parsed = lua_parser::var(&sequence).unwrap();
        let mut var = Var::Named(base);
        for property in properties {
            var = Var::PropertyAccess {
                from: Box::new(var),
                property,
            }
        }
        assert_eq!(var, parsed);
    }

    #[quickcheck]
    fn parse_single_member_lookup(base: Ident, expression: ArbitraryTokens<Expression>) {
        let mut tokens = vec![Token::Ident(base.clone()), Token::OpenSquareBracket];
        tokens.extend(expression.tokens);
        tokens.push(Token::CloseSquareBracket);
        let parsed = lua_parser::var(&tokens).unwrap();
        assert_eq!(
            parsed,
            Var::MemberLookup {
                from: Box::new(Var::Named(base)),
                value: Box::new(expression.expected)
            }
        );
    }

    #[quickcheck]
    fn parse_arbitrary_member_lookup(base: Ident, expressions: Vec<ArbitraryTokens<Expression>>) {
        let mut sequence = Vec::with_capacity(expressions.len() * 3 + 1);
        sequence.push(Token::Ident(base.clone()));
        sequence.extend(expressions.iter().flat_map(|expression| {
            let mut tokens = Vec::with_capacity(expression.tokens.len() + 3);
            tokens.push(Token::OpenSquareBracket);
            tokens.extend_from_slice(&expression.tokens);
            tokens.push(Token::CloseSquareBracket);
            tokens
        }));
        let parsed = lua_parser::var(&sequence).unwrap();
        let mut var = Var::Named(base);
        for expression in expressions {
            var = Var::MemberLookup {
                from: Box::new(var),
                value: Box::new(expression.expected),
            }
        }
        assert_eq!(var, parsed);
    }

    #[quickcheck]
    fn parse_arbitrary_var(ArbitraryTokens { tokens, expected }: ArbitraryTokens<Var>) {
        let parsed = lua_parser::var(&tokens).unwrap();
        assert_eq!(parsed, expected);
    }

    #[test]
    fn operator_precedence_1() {
        /*
        (
            (
                (
                    (
                        (
                            not (
                                (1)^(2)
                            )
                        ) * (
                            - (3)
                        )
                    ) + (
                        (4) / (5)
                    )
                ) .. (
                    (6) - (7)
                )
            ) < (8)
        ) and (
            (9) > (10)
        ) or (
            (11) == (12)
        )
        */
        static TEXT: &str = indoc! {"
            not 1^2 * - 3 + 4 / 5 .. 6 - 7 < 8 and 9 > 10 or 11 == 12
        "};
        let tokens: Vec<_> = Token::lexer(TEXT).collect();
        let expression = lua_parser::expression(&tokens).unwrap();
        insta::assert_debug_snapshot!(expression);
    }

    #[test]
    fn operator_precedence_2() {
        /*
        (
            (1) <= (2)
        ) and (
            (
                (3) + (
                    (
                        not (4)
                    ) * (
                        (5) ^ (6)
                    )
                )
            ) >= (
                (7) - (
                    (
                        - (8)
                    ) / (9)
                )
            )
        ) or (
            (10) ~= (
                (11) .. (12)
            )
        )
        */
        static TEXT: &str = indoc! {"
            1 <= 2 and 3 + not 4 * 5 ^ 6 >= 7 - - 8 / 9 or 10 ~= 11 .. 12
        "};
        let tokens: Vec<_> = Token::lexer(TEXT).collect();
        let expression = lua_parser::expression(&tokens).unwrap();
        insta::assert_debug_snapshot!(expression);
    }

    #[test]
    fn operator_precedence_3() {
        /*
        (
            (
                (1) <= (
                    (
                        (2) and (3)
                    ) + (
                        not (
                            (
                                (4) * (5)
                            )
                        ) ^ (6)
                    )
                )
            ) >= (
                (
                    (7) - (
                        (
                            - (8)
                        ) / (
                            (
                                (9) or (10)
                            ) ~= (11)
                        )
                    )
                ) .. (12)
            )
        )
        */
        static TEXT: &str = indoc! {"
            1 <= ((2 and 3) + not (4 * 5)) ^ 6 >= 7 - - 8 / ((9 or 10) ~= 11) .. 12
        "};
        let tokens: Vec<_> = Token::lexer(TEXT).collect();
        let expression = lua_parser::expression(&tokens).unwrap();
        insta::assert_debug_snapshot!(expression);
    }
}
