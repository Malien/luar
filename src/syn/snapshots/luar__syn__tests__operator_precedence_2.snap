---
source: src/syn/mod.rs
expression: expression
---
BinaryOperator {
    lhs: BinaryOperator {
        lhs: BinaryOperator {
            lhs: Number(
                NumberLiteral(
                    1.0,
                ),
            ),
            op: LessOrEquals,
            rhs: Number(
                NumberLiteral(
                    2.0,
                ),
            ),
        },
        op: And,
        rhs: BinaryOperator {
            lhs: BinaryOperator {
                lhs: Number(
                    NumberLiteral(
                        3.0,
                    ),
                ),
                op: Plus,
                rhs: BinaryOperator {
                    lhs: UnaryOperator {
                        op: Not,
                        exp: Number(
                            NumberLiteral(
                                4.0,
                            ),
                        ),
                    },
                    op: Mul,
                    rhs: BinaryOperator {
                        lhs: Number(
                            NumberLiteral(
                                5.0,
                            ),
                        ),
                        op: Exp,
                        rhs: Number(
                            NumberLiteral(
                                6.0,
                            ),
                        ),
                    },
                },
            },
            op: GreaterOrEquals,
            rhs: BinaryOperator {
                lhs: Number(
                    NumberLiteral(
                        7.0,
                    ),
                ),
                op: Minus,
                rhs: BinaryOperator {
                    lhs: UnaryOperator {
                        op: Minus,
                        exp: Number(
                            NumberLiteral(
                                8.0,
                            ),
                        ),
                    },
                    op: Div,
                    rhs: Number(
                        NumberLiteral(
                            9.0,
                        ),
                    ),
                },
            },
        },
    },
    op: Or,
    rhs: BinaryOperator {
        lhs: Number(
            NumberLiteral(
                10.0,
            ),
        ),
        op: NotEquals,
        rhs: BinaryOperator {
            lhs: Number(
                NumberLiteral(
                    11.0,
                ),
            ),
            op: Concat,
            rhs: Number(
                NumberLiteral(
                    12.0,
                ),
            ),
        },
    },
}
